src/memory/memory.c

```diff
@@ -150,13 +150,30 @@ static void _clean_chunk(uint8_t** chunk_bytes)
     uint8_t* __attribute__((__cleanup__(_clean_chunk))) __attribute__((unused)) var##_bytes = \
         (var).bytes;
 
+#ifndef TESTING
+static uint32_t _address_offset(memory_area_t area)
+{
+    switch (area) {
+    case MEMORY_AREA_SHARED_DATA:
+        return FLASH_SHARED_DATA_START;
+    case MEMORY_AREA_APP_DATA:
+        return FLASH_APPDATA_START;
+    case MEMORY_AREA_SMARTEEPROM:
+        return FLASH_SMARTEEPROM_START;
+    default:
+        Abort("unknown memory area");
+    }
+}
+#endif
+
 // chunk must have size CHUNK_SIZE. if chunk is NULL, the chunk is erased,
 // i.e. filled with 0xFF.
-static bool _write_to_address(uint32_t addr, uint8_t* chunk)
+static bool _write_to_address(memory_area_t area, uint32_t addr, uint8_t* chunk)
 {
 #ifdef TESTING
-    return memory_write_to_address_mock(addr, chunk);
+    return memory_write_to_address_mock(area, addr, chunk);
 #else
+    addr += _address_offset(area);
```

Comment:
forgot the comment: makes sense check for some addr overflow here? would want to avoid writing to some potentially sensitive memory chunks, even in theory.

Reply:
Which kind of overflow? You mean overflowing uint32?

Reply:
yeah, something like this: https://onlinegdb.com/EmYJiQDHa

Reply:
I renamed `addr` to `offset` and added an overflow check, PTAL. Note that unsigned overflow is well defined in C, so the check can be done after the addition.

Reply:
let's hope ARM's GCC adheres to that definition. fingers crossed.
j/k :) but here's a relevant real story: https://kqueue.org/blog/2013/09/17/cltq/

Reply:
I ran into undefined behavior with implicit numeric casts myself in the past :facepalm: grateful that Rust is here and we can abandon C forever :joy: